src/types/ErrorRate/utils.ts

```diff
@@ -0,0 +1,105 @@
+import { RateHealthConfig, RegexConfig, ToleranceConfig } from '../ServerConfig';
+import { serverConfig } from '../../config';
+import { ResponseDetail, Responses } from '../Graph';
+import { RequestHealth, RequestType } from '../Health';
+import { Rate, RequestTolerance } from './types';
+import { generateRateForTolerance } from './ErrorRate';
+import { generateRateForGraphTolerance } from './GraphEdgeStatus';
+
+export const emptyRate = (): Rate => {
+  return { requestRate: 0, errorRate: 0, errorRatio: 0 };
+};
+
+export const DEFAULTCONF = {
+  http: new RegExp('^[4|5]\\d\\d$'),
+  grpc: new RegExp('^[1-9]$|^1[0-6]$')
+};
+
+const requestsErrorRateCode = (requests: RequestType): number => {
+  const rate: Rate = emptyRate();
+  for (let [protocol, req] of Object.entries(requests)) {
+    for (let [code, value] of Object.entries(req)) {
+      rate.requestRate += value;
+      if (Object.keys(DEFAULTCONF).includes(protocol) && DEFAULTCONF[protocol].test(code)) {
+        rate.errorRate += value;
+      }
+    }
+  }
+  return rate.requestRate === 0 ? -1 : (rate.errorRate / rate.requestRate) * 100;
+};
+
+export const getErrorCodeRate = (requests: RequestHealth): { inbound: number; outbound: number } => {
+  return { inbound: requestsErrorRateCode(requests.inbound), outbound: requestsErrorRateCode(requests.outbound) };
+};
+
+export const checkExpr = (value: RegexConfig | undefined, testV: string): boolean => {
+  let reg = value;
+  if (!reg) {
+    return true;
+  }
+  if (typeof value === 'string') {
+    reg = new RegExp(value);
+  }
+  return (reg as RegExp).test(testV);
+};
+
+export const getConfig = (ns: string, name: string, kind: string): RateHealthConfig => {
+  if (serverConfig.healthConfig && serverConfig.healthConfig.rate) {
+    for (let rate of serverConfig.healthConfig.rate) {
+      if (checkExpr(rate.namespace, ns) && checkExpr(rate.name, name) && checkExpr(rate.kind, kind)) {
```

Comment:
Not for now, but a potential performance enhancement would be to use a helper map, to avoid to repeat the checkExpr() multiple times.

So, it can act as a cache of a key [ns, name, kind] with a rate.
Health config won't change unless there is a new server config, so, then that scenario would benefit of this.

To me it's fine to do it in a followup step, but at least let's leave a comment here with the idea.

Reply:
This is also something I raised previously, the high number of regex checks. In some situation I thought it is an overkill because there's a finite, low number of possible values (eg: direction either inbound or outbound or both).
BTW: do we have some measurement to track performance on the UI / frontend? (poke @FilipB @hhovsepy )

Reply:
Only performance tests I'm aware of are https://gitlab.cee.redhat.com/istio/kiali-qe/kiali-qe-utils/-/tree/master/perf/gatling
One is hitting K8sApis and second is hitting /api/namespaces/graph endpoint. We don't have any perf tests which would measure rendering times in browser. Adding new test for some REST endpoint would be easy if it helps here.